### ISSUE TITLE
error on code reload

### ISSUE THREAD

**Issue by PatrykKuniczak** [Jonghakseo/chrome-extension-boilerplate-react-vite#184]
**Describe the bug**
Sometimes when i add some more code, e.g 50 lines, then i click inside plugin in chrome, and then i have error(scren shot section):

And i have this maybe once per 10/15 reloads.

Good to know is, i don't close plugin, popup is all time open, and after edit code i click inside plugin.

Almost all times works well, but sometimes don't want to reload.

Maybe it's related to #157 

PS: When project grow up, then i one line change, may lead to an error.

**To Reproduce**
1. Clone and follow all steps for this repo reproduct.
2. Create e.g 1 custm component
3. Try to edit it, or paste several times in `Popup.tsx`

**Screenshots**
![Zrzut ekranu 2023-09-18 225639](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/9acfd0b3-a847-4155-a07b-746c598a7601)

For english: `Can't access file`.
File could be moved. Editted or removed.

PS: This info is misleading, because user see, he have some error in plugin, but everything works fine.

I think this console.warn which display that, should be e.g console.info or sth like that, what don't be displayed as error in extenstions section.

![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/4cdadff9-bcd8-41ed-8e1b-6d28c1dac592)

![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/a74ddee0-ab38-44d8-a8c2-0e8b13efb6f0)

**Desktop (please complete the following information):**
 - OS: Window 11
 - Browser: chrome latest
 - Node Version: 18.17.1

**Comment by Jonghakseo** [comment]
I'll take a look. :)

**Comment by PatrykKuniczak** [comment]
@Jonghakseo Any new info? :)

**Comment by Jonghakseo** [comment]
@PatrykKuniczak 

Yeah, I can't reproduce it very well. :(

#157 has been resolved, so I think it might have been resolved together, please check.

**Comment by PatrykKuniczak** [comment]
> @PatrykKuniczak
> 
> Yeah, I can't reproduce it very well. :(
> 
> #157 has been resolved, so I think it might have been resolved together, please check.

Ok, after work i take a look :)

**Comment by PatrykKuniczak** [comment]
@Jonghakseo It's still occur

![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/df055ee1-a442-4d79-affd-b1a55f556b2a)

As you can see, it's default view

And i'm copying several times this switch theme button.
I have all time open plugin in chrome.

And then after hmr build next time module, and it's finished.

Then i have this in chrome:
![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/4b909f5c-36ac-4f77-8f3f-25950a58cdfc)


After click outside(close plugin) and click again, to open, i have proper view:
![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/4aba7195-9729-4e4b-b734-32759977a652)

**Comment by Jonghakseo** [comment]
@PatrykKuniczak 

Ok... It's weired...
Can u reproduce this case on video?

https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/53500778/66b116ea-59b6-4c80-bd94-138317fbe706

https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/53500778/43d79dba-1147-4c6c-bac7-6e905518a306

**Comment by PatrykKuniczak** [comment]
As you can see i have it, still (I was pulled all new commits from main)

And for be sure, I turned it off all other plugins, but the bug still occur.


https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/b7f1c87f-9a58-4055-a989-3879eb62327e

![1](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/071ac2a5-6ace-4df0-8a72-2a9e3df2b386)
![2](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/42fe06ab-c73d-404c-9d49-8cdf0b226980)
![3](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/bd8bc3ad-72f6-42c4-8443-d1572e4c502d)
![4](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/cbda7c8f-783c-4e9d-8cc0-82f3d2a1d4f2)

**Comment by bambooom** [comment]
Hi, it there any progress on this issue? I also has the same error for twind
![Screenshot 2023-10-22 at 16 31 14](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/14840170/ffab4ed6-2453-4f68-96f9-9088645e36fe)

**Comment by PatrykKuniczak** [comment]
For me it's now ocurr almost everytime

**Comment by PatrykKuniczak** [comment]
@Jonghakseo any moves forward?

**Comment by vms-code** [comment]
@bambooom @PatrykKuniczak the error you are referring to happens because the content page uses a shadow DOM and twind functions. The shadow DOM separates the environment from the main DOM, this breaks the CSS, helpful link:
[ShadowDom doc](https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM#encapsulation_from_css)

So the way they make the CSS work is by adding the following function on the src/pages/content/components/Demo/index.tsx page:
`attachTwindStyle(rootIntoShadow, shadowRoot);`

This function creates a new CSSStyleSheet(), then uses the twind function to add tailwind class rules to the stylesheet and finally uses the: `documentOrShadowRoot.adoptedStyleSheets = [sheet.target];` line which injects the CSS rules into the shadow DOM.

if you want to use custom css rules instead of twind rules you can change the code to this:

`
export function attachTwindStyle<T extends { adoptedStyleSheets: unknown }>(
  observedElement: Element,
  documentOrShadowRoot: T,
) {
  //const sheet = cssom(new CSSStyleSheet());
  //const tw = twind(config, sheet);
  //observe(tw, observedElement);
  
  const customSheet = new CSSStyleSheet();
  
  // this are the same rules you would put in a normal .css file
  const cssRules = " body { background: #000; } .myClass { color: #fff; } ..." 
  
  customSheet.replaceSync(cssRules);
  documentOrShadowRoot.adoptedStyleSheets = [customSheet];
}
`
other useful link: [replaceSync](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet/replaceSync)

**Comment by PatrykKuniczak** [comment]
> background: #000; } .myClass { color: #fff; } ...\" \n  \n  customSheet.replaceSync(cssRules

I will take a look on that, i remove twid from my project, cause i was using mui with styled components, solution for this repo is probably remove twid completely, and style with pure tailwind, if the twid don't want to work in content page

**Comment by Jonghakseo** [comment]
> @Jonghakseo any moves forward?

Nope.

**Comment by Jonghakseo** [comment]
> > background: #000; } .myClass { color: #fff; } ..." \n  \n  customSheet.replaceSync(cssRules
> 
> I will take a look on that, i remove twid from my project, cause i was using mui with styled components, solution for this repo is probably remove twid completely, and style with pure tailwind, if the twid don't want to work in content page

I agree. I'm going to pull out the twind solution this weekend, and I'm thinking of configuring the solution that applies tailwind to be opt-in as a separate application method. I don't see any reason to put unnecessary code in for users who aren't going to use tailwind anyway.

**Comment by Jonghakseo** [comment]
- https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/pull/244

I don't think the removal of TWIND will cause the warning, but the reload issue that was initially raised as an issue is hard to reproduce.

Due to my situation, it is difficult to access the TWIND PC development environment... I hope someone can help :(

**Comment by PatrykKuniczak** [comment]
@Jonghakseo 
If i do

```ts
  socket.onclose = () => {
    console.warn(
      `Reload server disconnected.\nPlease check if the WebSocket server is running properly on ${LOCAL_RELOAD_SOCKET_URL}. This feature detects changes in the code and helps the browser to reload the extension or refresh the current tab.`,
    );
  };
```

in ```initReloadClient.ts```

then i can see an error, but it's sth strange:

![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/c0b2b7c3-4e01-4699-a8ce-d7452e773563)

Also still this error ocurr (more like info):
![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/b63377d1-a3fe-40a7-8e2c-af88d83991a5)

This info is cause the socket is disconnected for a second, and lose refrerence or STH like that, then can't woke up after all modules are reloaded, i try to solve this, but i don't create this hot module, and i must see how it's built.

**Comment by PatrykKuniczak** [comment]
@Jonghakseo Probably we have 2 separate issue, but maybe it's cause after i working on #233 i don't remove node_modules and there's some conflict, i remove entirely node_modules and dist and i will try to handle this again, it's really strange bug xD

**Comment by PatrykKuniczak** [comment]
@Jonghakseo It's sth with that i think:

![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/9369ade1-5c21-47c6-9296-ded56ec45d50)

But its really hard to debug

**Comment by PatrykKuniczak** [comment]
![image](https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/assets/64608510/2920da68-8680-45fd-90cf-1e928ee65a33)

I don't see it before, it's probably after add pnpm to the project, but i'm not sure.

But before we can use this template with trouble for time to time, but not it's completely useless, break on each/one to 2/3 code change, and not like before, only popup stop, and i must open it again, now the entire server is down, and i must reload it, that's big issue.

**Comment by PatrykKuniczak** [comment]
@Jonghakseo on your PC it works properly?

**Comment by PatrykKuniczak** [comment]
@Jonghakseo It's sth with view.ts

when on 

```ts
function reload(): void {
    pendingReload = false;
    window.location.reload();
  }
```

reload window, then 

``ts
 function reloadWhenTabIsVisible(): void {
    !document.hidden && pendingReload && reload();
  }

  document.addEventListener('visibilitychange', reloadWhenTabIsVisible);
  ```
  
  Handle visibility change, and run it again, and sth like loop is played here, maybe i'm wrong, but for me it's a possibly that, i must try to reproduce this, i will update here if i'll find sth new.

**Comment by Jonghakseo** [comment]
I found a similar issue in your error logs, but I don't know the cause yet.

https://github.com/nodejs/node/issues/31702

**Comment by Jonghakseo** [comment]
@PatrykKuniczak 

Since it's a permissions issue anyway, how about assigning higher permissions to that folder?

As far as I can tell from the error log, it seems to be a permissions issue, which is necessary to remove existing files during the build process.

**Comment by Jonghakseo** [comment]
> @Jonghakseo on your PC it works properly?

I don't have window PC... 😞

**Comment by PatrykKuniczak** [comment]
@Jonghakseo Maybe you can try on virtual env, sandbox or sth like that

**Comment by PatrykKuniczak** [comment]
> @PatrykKuniczak 
> 
> Since it's a permissions issue anyway, how about assigning higher permissions to that folder?
> 
> As far as I can tell from the error log, it seems to be a permissions issue, which is necessary to remove existing files during the build process.

You have any idea how to do it?

**Comment by Jonghakseo** [comment]
> @Jonghakseo Maybe you can try on virtual env, sandbox or sth like that

You're right! I will try it.

**Comment by PatrykKuniczak** [comment]
@Jonghakseo I see apple have sth like that: https://support.apple.com/pl-pl/guide/mac-help/mh11850/mac

Install windows is better option, if u can, cause sometimes on virtual env sth can work in differen way, but after all virtual env is better than nothing :)

**Comment by PatrykKuniczak** [comment]
@Jonghakseo I solve this part with permission,

```ts
chokidar.watch('src', { ignorePermissionErrors: true })
```

and also for 'dist', this error with crashing views scripts also ocurr..

If u have any idea feel free to push #256

**Comment by PatrykKuniczak** [comment]
@Jonghakseo Theres issue with this error:
https://github.com/Jonghakseo/chrome-extension-boilerplate-react-vite/issues/100

this race condition is the key of this problem.

I see the solution, cause delay isn't good solution, cause we can't be confident, how much it can take to reload entire code.

We must reformat that, and when 1 'watch' process works, don't run next for avoid race condition, when dev make change during rebuilding, this changes will be taken after the previouse process is done or kill the process when dev make change during rebuilding, sth like that nestjs have implemented.

I think in that case, we should listent on both sides od websockets, when client side, send complete message, then on server side, unlock next rebuild, this could work, but performance could be weak, for better performance we could, kill this process and run next one.

Then we can remove entirely delay.

Maybe if this don't work, we can use chokidar events to handle that, e.g. on 'unlink' block next process and on 'ready' check for new changes.

If you have any idea, let share with me :)

I will take a look for that tomorrow.

**Comment by Jonghakseo** [comment]
> @Jonghakseo Theres issue with this error: #100
> 
> this race condition is the key of this problem.
> 
> I see the solution, cause delay isn't good solution, cause we can't be confident, how much it can take to reload entire code.
> 
> We must reformat that, and when 1 'watch' process works, don't run next for avoid race condition, when dev make change during rebuilding, this changes will be taken after the previouse process is done or kill the process when dev make change during rebuilding, sth like that nestjs have implemented.
> 
> I think in that case, we should listent on both sides od websockets, when client side, send complete message, then on server side, unlock next rebuild, this could work, but performance could be weak, for better performance we could, kill this process and run next one.
> 
> Then we can remove entirely delay.
> 
> Maybe if this don't work, we can use chokidar events to handle that, e.g. on 'unlink' block next process and on 'ready' check for new changes.
> 
> If you have any idea, let share with me :)
> 
> I will take a look for that tomorrow.
> 
> PS: If any of this solution will work, then we can remove closing ws between rebuildings, cause this is also problematic.
> 
> I think we should establish 1 connection on start script and this WSS should run till, we or error shut down script(read node server ;))

@PatrykKuniczak 

Cool! I think the biggest problem is that the dist folder is detected by the file system.
I had an idea to send the build completion message from a custom plugin that is closely related to the build lifecycle, I'll try it.
You're really great.
